File: angular-app/src/app/core/application/decorator/helpers/interpreter/decorator-interpreter.ts
import { IOutputPresenter } from '../../../../presentation/interfaces/i-output-presenter';
import { UserAction } from '../../../enums/user-action';
import { DomainConstructor } from '../../../strategies/abstracts/base-application-strategy';
import { DecoratorTextHelper } from '../decorator-text-helper';
import { BaseHelperStrategy } from '../strategies/abstracts/base-helper-strategy';

export class DecoratorInterpreter {
  private readonly strategies: BaseHelperStrategy[];
  private readonly expectedMap: Partial<Record<UserAction, string>> = {
    [UserAction.LIST_MODELS]: 'listmodels',
    [UserAction.EXIT]: 'exit',
    [UserAction.CLEAR]: 'cls',
  };

  constructor(strategies: Iterable<BaseHelperStrategy>) {
    this.strategies = Array.from(strategies).sort((a, b) => a['priority'] - b['priority']);
  }

  interpretUserInputWithFeedback(
    userInput: string,
    presenter: IOutputPresenter
  ): [UserAction, string | DomainConstructor | null] {
    const normalized = DecoratorTextHelper.normalizeText(userInput);
    const tokens = normalized.split(' ');

    const [action, value] = this._interpret(normalized, tokens);

    const expected = this.expectedMap[action as UserAction];
    const compactInput = normalized.replace(/ /g, '');

    if (expected && compactInput !== expected.replace(/ /g, '')) {
      presenter.showInterpretedInput(userInput, expected);
    }

    return [action, value];
  }

  private _interpret(normalized: string, tokens: string[]): [UserAction, string | DomainConstructor | null] {
    const candidate = normalized.replace(/ /g, '');

    for (const strategy of this.strategies) {
      if (strategy.canHandle(candidate, tokens)) {
        return strategy.handle(candidate, tokens);
      }
    }

    return [UserAction.MESSAGE, normalized];
  }
}

File: python-app/application/decorator/interpreter/decorator_interpreter.py
from collections.abc import Iterable
from typing import Any
from application.enums.user_action import UserAction
from application.decorator.helpers.decorator_text_helper import DecoratorTextHelper
from application.decorator.strategies.abstracts.base_helper_strategy import BaseHelperStrategy
from presentation.presenters.i_output_presenter import IOutputPresenter

class DecoratorInterpreter:

    def __init__(self, strategies: Iterable[BaseHelperStrategy]):
        self._strategies = list(strategies)
        self._strategies.sort(key=lambda s: getattr(s, "_priority", 0))

        self._expected_map = {
            UserAction.LIST_MODELS: "list models",
            UserAction.EXIT: "exit",
            UserAction.CLEAR: "cls",
        }

    def interpret_user_input_with_feedback(
        self,
        user_input: str,
        presenter: IOutputPresenter
    ) -> tuple[UserAction, Any | None]:

        normalized = DecoratorTextHelper.normalize_text(user_input)
        tokens = normalized.split()

        action, value = self._interpret(normalized, tokens)

        expected = self._expected_map.get(action)
        compact_input = normalized.replace(" ", "")

        if expected and compact_input != expected.replace(" ", ""):
            presenter.show_interpreted_input(user_input, expected)

        return action, value

    def _interpret(self,
        normalized: str,
        tokens: list[str]) -> tuple[UserAction, Any | None]:

        candidate = normalized.replace(" ", "")

        for strategy in self._strategies:
            if strategy.can_handle(candidate, tokens):
                return strategy.handle(candidate, tokens)

        return UserAction.MESSAGE, normalized
